### PR TITLE
ENH: TIMED GUVI

### PR DESCRIPTION
**Pull request by aburrell**
# Description

Adds a TIMED GUVI instrument capable of handling the EDR Auroral data.  Builds on #147 and so must be merged after that pull request.  Additionally, improves testing for the platform modules and generalized support for Imager data processed by JHU APL.  Streamlines previous TIMED GUVI loading, ensured all metadata was loaded appropriately, and added a custom flag for time variables, to simplify potential indexing issues.

Although one file is renamed, I did not deprecate it because it was missing from the imports.

## Type of change

Please delete options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality
      to not work as expected)
- This change requires a documentation update

# How Has This Been Tested?

Added a new unit test.  Also:

```
import datetime as dt
import pysat
import pysatNASA

# Load raises logger warnings that "TIME" is not a dimension, as before
guvi = pysat.Instrument(inst_module=pysatNASA.instruments.timed_guvi, tag='edr-aur')
guvi.download(start=dt.datetime(2003, 1, 1))
guvi.load(date=dt.datetime(2003, 1, 1))
print(guvi.variables)
```

Yields:
```
['time', 'YEAR', 'DOY', 'LATITUDE', 'LONGITUDE', 'ALTITUDE', 'PIXELSIZE_GEOMAGNETIC_LATITUDE', 'PIXELSIZE_GEOMAGNETIC_LONGITUDE', 'NORTH_DATA', 'SOUTH_DATA', 'NORTH_SWATH_BOUNDARY', 'SOUTH_SWATH_BOUNDARY', 'SOUTH_GEOMAGNETIC_LATITUDE', 'SOUTH_GEOMAGNETIC_LONGITUDE', 'SOUTH_MAGNETIC_LOCAL_TIME', 'SOUTH_GEOGRAPHIC_LATITUDE', 'SOUTH_GEOGRAPHIC_LONGITUDE', 'MODEL_SOUTH_GEOMAGNETIC_LATITUDE', 'MODEL_SOUTH_GEOMAGNETIC_LATITUDE_ERROR', 'MODEL_SOUTH_GEOMAGNETIC_LONGITUDE', 'MODEL_SOUTH_MAGNETIC_LOCAL_TIME', 'MODEL_SOUTH_GEOGRAPHIC_LATITUDE', 'MODEL_SOUTH_GEOGRAPHIC_LONGITUDE', 'SOUTH_POLAR_GEOMAGNETIC_LATITUDE', 'SOUTH_POLAR_GEOMAGNETIC_LONGITUDE', 'SOUTH_POLAR_MAGNETIC_LOCAL_TIME', 'SOUTH_POLAR_GEOGRAPHIC_LATITUDE', 'SOUTH_POLAR_GEOGRAPHIC_LONGITUDE', 'MODEL_SOUTH_POLAR_GEOMAGNETIC_LATITUDE', 'MODEL_SOUTH_POLAR_GEOMAGNETIC_LONGITUDE', 'MODEL_SOUTH_POLAR_MAGNETIC_LOCAL_TIME', 'MODEL_SOUTH_POLAR_GEOGRAPHIC_LATITUDE', 'MODEL_SOUTH_POLAR_GEOGRAPHIC_LONGITUDE', 'SOUTH_DEFAULT_EQUATORWARD_MODEL', 'SOUTH_DATA_QUALITY', 'UT_N', 'NORTH_GEOMAGNETIC_LATITUDE', 'NORTH_GEOMAGNETIC_LONGITUDE', 'NORTH_MAGNETIC_LOCAL_TIME', 'NORTH_GEOGRAPHIC_LATITUDE', 'NORTH_GEOGRAPHIC_LONGITUDE', 'MODEL_NORTH_GEOMAGNETIC_LATITUDE', 'MODEL_NORTH_GEOMAGNETIC_LONGITUDE', 'MODEL_NORTH_GEOMAGNETIC_LATITUDE_ERROR', 'MODEL_NORTH_MAGNETIC_LOCAL_TIME', 'MODEL_NORTH_GEOGRAPHIC_LATITUDE', 'MODEL_NORTH_GEOGRAPHIC_LONGITUDE', 'NORTH_POLAR_GEOMAGNETIC_LATITUDE', 'NORTH_POLAR_GEOMAGNETIC_LONGITUDE', 'NORTH_POLAR_MAGNETIC_LOCAL_TIME', 'NORTH_POLAR_GEOGRAPHIC_LATITUDE', 'NORTH_POLAR_GEOGRAPHIC_LONGITUDE', 'MODEL_NORTH_POLAR_GEOMAGNETIC_LATITUDE', 'MODEL_NORTH_POLAR_GEOMAGNETIC_LONGITUDE', 'MODEL_NORTH_POLAR_MAGNETIC_LOCAL_TIME', 'MODEL_NORTH_POLAR_GEOGRAPHIC_LATITUDE', 'MODEL_NORTH_POLAR_GEOGRAPHIC_LONGITUDE', 'NORTH_DEFAULT_EQUATORWARD_MODEL', 'NORTH_DATA_QUALITY', 'UT_S', 'ELECTRON_MEAN_NORTH_ENERGY_MAP', 'ELECTRON_MEAN_SOUTH_ENERGY_MAP', 'ENERGY_FLUX_NORTH_MAP', 'ENERGY_FLUX_SOUTH_MAP', 'ELECTRON_FLUX_NORTH_BOUNDARY_MAP', 'ELECTRON_FLUX_SOUTH_BOUNDARY_MAP', 'LATITUDE_GEOMAGNETIC_GRID_MAP', 'LONGITUDE_GEOMAGNETIC_NORTH_GRID_MAP', 'LONGITUDE_GEOMAGNETIC_SOUTH_GRID_MAP', 'MLT_GRID_MAP', 'ELECTRON_ENERGY_FLUX_THRESHOLDS', 'HME_NORTH', 'NME_NORTH', 'HME_SOUTH', 'NME_SOUTH', 'PROTON_FLAG_NORTH', 'PROTON_FLAG_SOUTH', 'HEMISPHERE_POWER_NORTH', 'HEMISPHERE_POWER_SOUTH', 'DISK_RADIANCEDATA_INTENSITY_NORTH', 'DISK_DQI_NORTH', 'DISK_RADIANCEDATA_INTENSITY_SOUTH', 'DISK_DQI_SOUTH', 'DATA_QUALITY_GLOBAL', 'AUR_ARC_RADIANCE_NORTH', 'AUR_ARC_RADIANCE_SOUTH', 'AUR_ARC_GROUP_MAP_NORTH', 'AUR_ARC_GROUP_MAP_SOUTH', 'AUR_ARC_MEDIAN_RAD_NORTH', 'AUR_ARC_MEDIAN_RAD_SOUTH', 'AUR_ARC_BOX_VERTEX_NORTH', 'AUR_ARC_BOX_VERTEX_SOUTH', 'AUR_ARC_BOX_ORIENT_NORTH', 'AUR_ARC_BOX_ORIENT_SOUTH', 'MagneticFieldLines']
```

## Test Configuration
* Operating system: OS X BIg Sur
* Version number: Python 3.8
* Any details about your local setup that are relevant: develop branch of pysat

# Checklist:

- [x] Make sure you are merging into the ``develop`` (not ``main``) branch
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] Add a note to ``CHANGELOG.md``, summarizing the changes
- [x] Update zenodo.json file for new code contributors

### CODE REVIEW

File: pysatNASA/instruments/methods/jhuapl.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    -------
+    data : pds.DataFrame or xr.Dataset
+        Data to be assigned to the pysat.Instrument.data object.
+    mdata : pysat.Meta
+        Pysat Meta data for each data variable.
+
+    Note
+    ----
+    Logger warning 'Epoch label: TIME is not a dimension.' is raised due to
+    the data format and pysat file expectations.
```
Comment (aburrell) [inline pysatNASA/instruments/methods/jhuapl.py:37]: @jklenzing added note in response to feedback from #147

**Review by aburrell — COMMENTED** [review]
Need to decide if this should wait on #138 before adapting/merging this pull request.

**Comment by jklenzing** [comment]
@aburrell, flake8 is erroring on an unbalanced bracket.  It looks like there were some additional style issues in the merge. I've fixed it locally, running the tests now to see if all works as expected.  Checking in to see if you fixed this locally as well before I push.

File: pysatNASA/instruments/timed_guvi.py [hunk 2]
```diff
@@ -47,108 +69,74 @@
...
+multi_file_day = True  # This is only true for EDR-AUR
 
 # ----------------------------------------------------------------------------
 # Instrument test attributes
 
-_test_dates = {
-    'high_res': {tag: dt.datetime(2005, 6, 28) for tag in tags.keys()},
-    'low_res': {tag: dt.datetime(2005, 6, 28) for tag in tags.keys()}}
+_test_dates = {iid: {tag: dt.datetime(2005, 6, 28) for tag in inst_ids[iid]
+               for iid in inst_ids.keys()}
```
Comment (jklenzing) [inline pysatNASA/instruments/timed_guvi.py:95]: This is the source of the SyntaxError, which points to the following lines as the problem.

Comment (aburrell) [reply pysatNASA/instruments/timed_guvi.py:95]: I don't have more time to work on this today, go ahead and push if you made the changes already 🦦

File: pysatNASA/instruments/timed_guvi.py [hunk 3]
```diff
@@ -1,11 +1,31 @@
...
 
-Downloads data from the NASA Coordinated Data
-Analysis Web (CDAWeb).
+Supports the Global UltraViolet Imager (GUVI) instrument on the Thermosphere
+Ionosphere Mesosphere Energetics Dynamics (TIMED) satellite data from the
+NASA Coordinated Data Analysis Web (CDAWeb).
 
-Supports two options for loading that may be
-specified at instantiation.
+From JHU APL (L2 Data):
```
Comment (aburrell) [inline pysatNASA/instruments/timed_guvi.py:8]:
```suggestion
From JHU APL:
```

File: pysatNASA/instruments/timed_guvi.py [hunk 4]
```diff
@@ -191,166 +179,163 @@ def load(fnames, tag='', inst_id=''):
...
+                    days=float(dataset['DOY_%s' % tag].data[i] - 1),
+                    seconds=float(dataset['TIME_%s' % tag].data[i]))
+                dts.append(idt)
+            return dts
+
+        # Dimensions for time variables
+        # night/day along, cross and time variables are within
+        #    imaging low_res/high_res and spectrograph high_res/low_res
+        # imaging high_res also has dims DayAur
+        # spectrograph low_res also has dims GAIMDay, GAIMNight
```
Comment (aburrell) [inline pysatNASA/instruments/timed_guvi.py:201]: Recommend fixing spacing and punctuation.

File: pysatNASA/instruments/timed_guvi.py [hunk 5]
```diff
@@ -191,166 +179,163 @@ def load(fnames, tag='', inst_id=''):
...
+            data = data.swap_dims({"nAlong": "time"})
+            if 'sdr-imag' in tag:
+                data = data.swap_dims({"nAlongDayAur": "time_auroral"})
+            elif 'sdr-spect' in tag:
+                if 'low' in inst_id:
+                    data = data.swap_dims({"nAlongGAIMDay": "time_gaim_day",
+                                           "nAlongGAIMNight": "time_gaim_night"})
+
+            # Update time variables
+            # night_dts and day_dts are the same temporal array
```
Comment (aburrell) [inline pysatNASA/instruments/timed_guvi.py:268]: Update comment formatting

File: pysatNASA/instruments/timed_guvi.py [hunk 6]
```diff
@@ -191,166 +179,163 @@ def load(fnames, tag='', inst_id=''):
...
-
-        # 'nCross' dimension only in imaging
-        if 'sdr-imag' in tag:
-
-            if data.nCrossDay.size != data.nCrossNight.size:
-                raise Exception("nCrossDay/Night have different dimensions")
+            coords = coords + ['PIERCEPOINT_NIGHT_ZENITH_ANGLE',
+                               'PIERCEPOINT_NIGHT_SAZIMUTH',
+                               'PIERCEPOINT_DAY_ZENITH_ANGLE',
+                               'PIERCEPOINT_DAY_SAZIMUTH']
```
Comment (aburrell) [inline pysatNASA/instruments/timed_guvi.py:320]: Recommend changing style to use `extend`

File: pysatNASA/instruments/timed_guvi.py [hunk 7]
```diff
@@ -191,166 +179,163 @@ def load(fnames, tag='', inst_id=''):
...
-        coords = {"nchan": ["121.6nm", "130.4nm", "135.6nm",
-                            "LBHshort", "LBHlong", "?"]}
-
-    data = data.assign_coords(coords=coords)
-
-    # Sort
-    data = data.sortby("time")
+        data = data.assign_coords(coords=coords)
+
+        # Sort
```
Comment (aburrell) [inline pysatNASA/instruments/timed_guvi.py:338]: Make the comment a sentence

**Review by aburrell — COMMENTED** [review]
Some style things that I need to fix.

File: pysatNASA/instruments/timed_guvi.py (hunk 2 above)
Comment (jklenzing) [reply pysatNASA/instruments/timed_guvi.py]: I've pushed the changes.  flake 8 passes locally, but there's still a download error.  Not sure if this is unique to me or if we'll see it on github actions too

File: pysatNASA/instruments/methods/jhuapl.py [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+import datetime as dt
+import numpy as np
+import pandas as pds
+import xarray as xr
+
+from pysat.utils.io import load_netcdf
+
+
+def build_dtimes(data, var, epoch=None, epoch_var='time'):
+    """Build datetime objects from standard JPL time variables.
```
Comment (aburrell) [inline pysatNASA/instruments/methods/jhuapl.py:13]:
```suggestion
    """Build datetime objects from standard JHU APL time variables.
```

**Review by aburrell — COMMENTED** [review]
Waiting to commit fix until after additional feedback to reduce test runs.

**Comment by aburrell** [comment]
@jklenzing the load options don't seem to be resulting in the expected coverage.  Is it not doing what I think it's doing?

**Comment by jklenzing** [comment]
> @jklenzing the load options don't seem to be resulting in the expected coverage. Is it not doing what I think it's doing?

The list of tests with optional kwargs looks right for the new guvi data.  not sure what is missing.

File: pysatNASA/instruments/timed_guvi.py [hunk 9]
```diff
@@ -38,121 +60,86 @@
...
 # Instrument test attributes
 
-_test_dates = {
-    'high_res': {tag: dt.datetime(2005, 6, 28) for tag in tags.keys()},
-    'low_res': {tag: dt.datetime(2005, 6, 28) for tag in tags.keys()}}
+_test_dates = {iid: {tag: dt.datetime(2005, 6, 28) for tag in inst_ids[iid]}
+               for iid in inst_ids.keys()}
+_test_load_opt = {iid: {tag: [{'combine_times': True}, {'combine_times': False}]
+                        for tag in inst_ids[iid]} for iid in ['high_res',
+                                                              'low_res']}
```
Comment (jklenzing) [inline pysatNASA/instruments/timed_guvi.py:93]:
```suggestion
_test_load_opt = {iid: {tag: {'combine_times': True}
                        for tag in inst_ids[iid]} for iid in ['high_res',
                                                              'low_res']}
```

There seems to be a bug on the pysat side.  It's doubling up on combine_times=False.  In general, I've only put the non-default option up into `_test_load_opt`.  This should fix the coverage issue.

**Comment by jklenzing** [comment]
Documented the pysat bug at https://github.com/pysat/pysat/issues/1097

**Review by jklenzing — APPROVED** [review]

**Pull request merged**